store/multiversion/data_structures.go

```diff
@@ -0,0 +1,160 @@
+package multiversion
+
+import (
+	"sync"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/google/btree"
+)
+
+const (
+	// The approximate number of items and children per B-tree node. Tuned with benchmarks.
+	multiVersionBTreeDegree = 2 // should be equivalent to a binary search tree TODO: benchmark this
+)
+
+type MultiVersionValue interface {
+	GetLatest() (value MultiVersionValueItem, found bool)
+	GetValueByIndex(index int) (value MultiVersionValueItem, found bool)
+	Set(index int, value []byte)
+	SetEstimate(index int)
+	Delete(index int)
+}
+
+type MultiVersionValueItem interface {
+	IsDeleted() bool
+	IsEstimate() bool
+	Value() []byte
+	Index() int
+}
+
+type multiVersionItem struct {
+	valueTree *btree.BTree // contains versions values written to this key
+	mtx       sync.RWMutex // manages read + write accesses
+}
+
+var _ MultiVersionValue = (*multiVersionItem)(nil)
+
+func NewMultiVersionItem() *multiVersionItem {
+	return &multiVersionItem{
+		valueTree: btree.New(multiVersionBTreeDegree),
+	}
+}
+
+// GetLatest returns the latest written value to the btree, and returns a boolean indicating whether it was found.
+//
+// A `nil` value along with `found=true` indicates a deletion that has occurred and the underlying parent store doesn't need to be hit.
+func (item *multiVersionItem) GetLatest() (MultiVersionValueItem, bool) {
+	item.mtx.RLock()
+	defer item.mtx.RUnlock()
+
+	bTreeItem := item.valueTree.Max()
+	if bTreeItem == nil {
+		return nil, false
+	}
+	valueItem := bTreeItem.(*valueItem)
+	return valueItem, true
+}
+
+// GetLatest returns the latest written value to the btree prior to the index passed in, and returns a boolean indicating whether it was found.
+//
+// A `nil` value along with `found=true` indicates a deletion that has occurred and the underlying parent store doesn't need to be hit.
+func (item *multiVersionItem) GetValueByIndex(index int) (MultiVersionValueItem, bool) {
```

Comment:
nit: naming is a bit misleading here since we won't get the value at `index` even if there exists one